### PR TITLE
#3277 Implements threshold for HighlightLayer introducing threshold p…

### PR DESCRIPTION
**Pull request by PixelsCommander**
#3277 Implements threshold for HighlightLayer introducing threshold property on HighlightLayerOptions. Code for demoing this - https://playground.babylonjs.com/#K7UNC7

### CODE REVIEW

**Comment by PixelsCommander** [comment]
screen shot 2017-12-11 at 2 01 21 pm

File: src/Layer/babylon.highlightlayer.ts [hunk 1]
```diff
@@ -1,4 +1,19 @@
 ﻿module BABYLON {
+    BABYLON.Effect.ShadersStore["highlightLayerThresholdFragmentShader"] = "\r\n" +
```
Comment (deltakosh) [inline src/Layer/babylon.highlightlayer.ts]: shaders have to be stored as file in /src/shaders

File: src/Layer/babylon.highlightlayer.ts [hunk 2]
```diff
@@ -400,14 +421,26 @@
                 });
             }
 
+            const postProcesses = [this._downSamplePostprocess, this._horizontalBlurPostprocess, this._verticalBlurPostprocess];
+
+            if (this._options.threshold) {
+                const threshold = this._options.threshold;
+                this._thresholdPostProcess = new BABYLON.PostProcess("thresold", "highlightLayerThreshold", ["screenSize", "threshold"], null, 0.25, null, BABYLON.Texture.BILINEAR_SAMPLINGMODE, this._scene.getEngine());
```
Comment (deltakosh) [inline src/Layer/babylon.highlightlayer.ts]: must be disposed

Comment (PixelsCommander) [reply src/Layer/babylon.highlightlayer.ts]: Fixed

**Comment by PixelsCommander** [comment]
@deltakosh all your comments were fixed

**Comment by deltakosh** [comment]
Perfect!! Will wait for 3.1 release now :)

File: src/Layer/babylon.highlightlayer.ts [hunk 3]
```diff
@@ -56,6 +56,11 @@
          * The camera attached to the layer.
          */
         camera: Nullable<Camera>;
+
+        /**
+         * Threshold for cutting glow.
+         */
+        threshold: number;
```
Comment (deltakosh) [inline src/Layer/babylon.highlightlayer.ts]: should be threshold?: number

File: src/Layer/babylon.highlightlayer.ts [hunk 4]
```diff
@@ -400,14 +406,26 @@
                 });
             }
 
+            const postProcesses = [this._downSamplePostprocess, this._horizontalBlurPostprocess, this._verticalBlurPostprocess];
+
+            if (this._options.threshold) {
+                const threshold = this._options.threshold;
+                this._thresholdPostProcess = new BABYLON.PostProcess("thresold", "highlightLayerThreshold", ["screenSize", "threshold"], null, 0.25, null, BABYLON.Texture.BILINEAR_SAMPLINGMODE, this._scene.getEngine());
```
Comment (deltakosh) [inline src/Layer/babylon.highlightlayer.ts]: PP name should be threshold ;)

**Comment by deltakosh** [comment]
Up?

**Comment by PixelsCommander** [comment]
@deltakosh all s fixed

File: src/Layer/babylon.highlightlayer.ts [hunk 5]
```diff
@@ -879,6 +897,7 @@
             this._downSamplePostprocess.dispose();
             this._horizontalBlurPostprocess.dispose();
             this._verticalBlurPostprocess.dispose();
+            this._thresholdPostProcess.dispose();
```
Comment (deltakosh) [inline src/Layer/babylon.highlightlayer.ts]: Need a if here as the threshold Pp can be undefined

Comment (PixelsCommander) [reply src/Layer/babylon.highlightlayer.ts]: Done

**Comment by deltakosh** [comment]
LGTM.
Waiting for @sebavan review

**Comment by sebavan** [comment]
Could the threshold post process logic be integrated in glowMapMerge under a define instead ?

This would save a full screen pass improving the perfromances.

**Comment by PixelsCommander** [comment]
@sebavan elaborate on "under define" approach please.

**Comment by deltakosh** [comment]
Not sure we can do it actually as the HL uses only default PP (like blur).

**Comment by sebavan** [comment]
Actually the shader https://github.com/BabylonJS/Babylon.js/blob/master/src/Shaders/glowMapMerge.fragment.fx#L10
is used to blit the blurred highlight information on top of the rendered scene.

So the code:

```
#ifdef THRESHOLD
float alpha = smoothstep(.0, threshold, baseColor.a);
baseColor.a = alpha;
baseColor.rgb= baseColor.rgb * alpha;
#endif
```

should be able to be inserted in the glowMapMerge shader to save one full pass.

The #ifdef THRESHOLD is the part I called Defines cause in Babylon THRESHOLD would need to be added to the defines list.

**Comment by deltakosh** [comment]
Ok makes sense..This is a great idea

**Comment by PixelsCommander** [comment]
@sebavan @deltakosh  And how THRESHOLD should be set from JS ?

**Comment by deltakosh** [comment]
You can add a new uniform here:
https://github.com/BabylonJS/Babylon.js/blob/master/src/Layer/babylon.highlightlayer.ts#L10

**Comment by deltakosh** [comment]
up?

**Comment by deltakosh** [comment]
Hello @PixelsCommander , do you have time to finish this PR?

**Comment by PixelsCommander** [comment]
Currently working on another issue. Will get back to stroke/threshold once done with it.

**Comment by deltakosh** [comment]
Sounds good

**Comment by PixelsCommander** [comment]
@deltakosh , @sebavan elaborate please on how should I define THRESHOLD to use with #ifdef

**Comment by deltakosh** [comment]
You also need to add threshold in the list of uniform.
Would you mind updating the PR to see current state?

**Comment by PixelsCommander** [comment]
I do not use uniforms anymore. No reason for stroke to be tunable, I found the number of 0.5 working pretty smooth then you control width of the stroke in the same way used for highlightLayer.

So the question is still the same - any hints on why defines are lost in Effect constructor?

**Comment by sebavan** [comment]
@PixelsCommander,  you added the stroke define on the wrong post process (glowBlurPostProcess instead of glowMapMergeEffect https://github.com/BabylonJS/Babylon.js/blob/master/src/Layer/babylon.highlightlayer.ts#L289)

Moving should solve your issue.

**Comment by PixelsCommander** [comment]
@deltakosh @sebavan check it now, the code for demoing https://playground.babylonjs.com/#K7UNC7

**Comment by sebavan** [comment]
Thank you, works nicely.

**Comment by deltakosh** [comment]
Please update the what's new as well to credit you

**Comment by PixelsCommander** [comment]
@deltakosh would this be a valid PR for this https://github.com/BabylonJS/Babylon.js/pull/3605 ?

**Pull request merged**